### ISSUE TITLE
inferior reproduction performance

### ISSUE THREAD

**Issue by yo3nglau** [ttlmh/Bridge-Prompt#2]
Hi~
Thank you for your amazing work !

When I conducted the action segmentation experiments on GTEA dataset with your official public code by myself, I found the reproduced results are inferior to ones that paper reported. I am wondering what I missed in the reproduction. Here allow me to depict several details of my trial.

I argue that each of 4 splits should be trained with Bridge-Prompt respectively so that their test splits will not participate in the training process, which might be not mentioned in the paper as well as the code README.md.

So I adjusted n_split in `gtea_ft.yaml` to utilize `train_split1/2/3/4_nf16_ol[2, 1, 0.5]_ds[1, 2, 4].npy`, the rest remained the same, to train the model separately. After that, each last model was regarded as the corresponding pretrained model of the specific split (by updating `pretrain` in `gtea_exfm.yaml` and `log_time` in `extract_frame_features.py`) for the purpose of extracting frame-wise features. (BTW, I present these two files I configured as `.txt` format at the end of this issue.)

In other words, each of 4 splits had its specified frame-wise features, which were further served as the model inputs of ASFormer. That is to say, formally I obtained four folder `gtea_vit_features_split1/2/3/4` to deposit extracted features for each individual split. 

To successfully match ASFormer input size `(features_dim, n_frames)`, I transposed the final feature presentations `(n_frames, features_dim)` and simply rectified the `features_dim=768` within main.py of ASFormer source code. As proposed in their paper, I train 120 epochs and evaluated the performance with the `120-epoch.model`.

The reproduced per and average results of 4 splits on GTEA dataset are shown below:
python eval.py --dataset=gtea --split=1
Acc: 81.8661  Edit: 88.776471  F1@10,25,50  [91.17647059 88.97058824 80.14705882]
python eval.py --dataset=gtea --split=2
Acc: 80.4668  Edit: 87.354703  F1@10,25,50  [87.94326241 85.81560284 77.30496454]
python eval.py --dataset=gtea --split=3
Acc: 81.1862  Edit: 92.991235  F1@10,25,50  [93.23308271 90.97744361 87.21804511]
python eval.py --dataset=gtea --split=4
Acc: 83.7258  Edit: 90.182133  F1@10,25,50  [94.73684211 93.98496241 83.45864662]
python eval.py --dataset=gtea --split=0
Acc: 81.8112  Edit: 89.826136  F1@10,25,50  [91.77241445290323, 89.93714927180278, 82.03217877296495]

Here I reveal the logs training Bridge-Prompt on GTEA from provided pretrained models `vit-16-32f.pt`:
Name | State | Notes | User | Runtime | lr | train_loss_acts | train_loss_all | train_loss_cnt | train_total_loss | Created
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
20220413_184847_clip_ucf_ViT-B/16_gtea | finished | gtea_split1 | yo3nglau | 15457 | 7.16E-12 | 0.56103516 | 0.10571289 | 0.59570313 | 1.26269531 | 2022-04-13T10:48:51.000Z
20220413_122249_clip_ucf_ViT-B/16_gtea | finished | gtea_split2 | yo3nglau | 15340 | 2.91E-12 | 0.20373535 | 0.02615356 | 0.47509766 | 0.70507813 | 2022-04-13T04:22:56.000Z
20220412_114851_clip_ucf_ViT-B/16_gtea | finished | gtea_split3 | yo3nglau | 22225 | 1.71E-11 | 0.25561523 | 0.07971191 | 0.06726074 | 0.40283203 | 2022-04-12T03:48:55.000Z
20220411_215453_clip_ucf_ViT-B/16_gtea | finished | gtea_split4 | yo3nglau | 16942 | 4.05E-12 | 0.3918457 | 0.05957031 | 0.13122559 | 0.58251953 | 2022-04-11T13:54:57.000Z

I am very confused and urgently look forward to receiving your patient guidance !

Best regards

[gtea_ft.txt](https://github.com/ttlmh/Bridge-Prompt/files/8486484/gtea_ft.txt)
[gtea_exfm.txt](https://github.com/ttlmh/Bridge-Prompt/files/8486488/gtea_exfm.txt)

**Comment by ttlmh** [comment]
Hi, thanks for your recognition of our work, and we also appreciate the detailed description you have given! Our training procedure for action segmentation is exactly the same as you have described.

**Comment by ttlmh** [comment]
Hi, thanks for your recognition of our work! Our training procedure for action segmentation is exactly the same as you have described, and we really appreciate the very detailed explanation you have given us (which can also be a useful complement to our repo).

We noticed that you trained the ASFormer for 120 epochs, and took the last epoch for evaluation. Here's the problem, we've found that Br-Prompt pre-trained ViT features (d=768) are more effective than the I3D features (d=2048) which were used in the original ASFormer settings. Thus, it may take fewer training epochs for ASFormer to converge. Then we decided to test the results on earlier epochs, and found that the performances could be higher. We're not sure if you used the same random settings to train the framework as we used for the paper, but you can test the model on earlier epochs to verify if there are better results.

**Comment by yo3nglau** [comment]
> Hi, thanks for your recognition of our work! Our training procedure for action segmentation is exactly the same as you have described, and we really appreciate the very detailed explanation you have given us (which can also be a useful complement to our repo).
> 
> We noticed that you trained the ASFormer for 120 epochs, and took the last epoch for evaluation. Here's the problem, we've found that Br-Prompt pre-trained ViT features (d=768) are more effective than the I3D features (d=2048) which were used in the original ASFormer settings. Thus, it may take fewer training epochs for ASFormer to converge. Then we decided to test the results on earlier epochs, and found that the performances could be higher. We're not sure if you used the same random settings to train the framework as we used for the paper, but you can test the model on earlier epochs to verify if there are better results.

Thank you for quick reply! I will make efforts based on your suggestion. If I obtain the best results with certain epoch model for each of 4 splits, but these 4 models might derive from different epochs, can I average these partial best results to acquire the last and of course best performance?

**Comment by yo3nglau** [comment]
> Hi, thanks for your recognition of our work! Our training procedure for action segmentation is exactly the same as you have described, and we really appreciate the very detailed explanation you have given us (which can also be a useful complement to our repo).
> 
> We noticed that you trained the ASFormer for 120 epochs, and took the last epoch for evaluation. Here's the problem, we've found that Br-Prompt pre-trained ViT features (d=768) are more effective than the I3D features (d=2048) which were used in the original ASFormer settings. Thus, it may take fewer training epochs for ASFormer to converge. Then we decided to test the results on earlier epochs, and found that the performances could be higher. We're not sure if you used the same random settings to train the framework as we used for the paper, but you can test the model on earlier epochs to verify if there are better results.

Thank you for quick reply! I immediately make efforts based on your suggestion. I met two puzzles. **(Question 1)** If I obtain the best results with certain epoch model for each of 4 splits, but these 4 models might derive from different epochs, can I average these partial best results to acquire the last and of course best performance? I assume that is allowed and the results are shown below:

- split1 epoch-110.model
Acc: 82.4479  Edit: 90.708101  F1@10,25,50  [93.38235294 91.17647059 81.61764706]

- split2 epoch-100.model
Acc: 80.7248  Edit: 87.533241  F1@10,25,50  [89.7810219  87.59124088 81.75182482]

- split3 epoch-100.model
Acc: 81.5399  Edit: 93.482953  F1@10,25,50  [95.09433962 94.33962264 89.05660377]

- split4 epoch-90.model
Acc: 83.7560  Edit: 91.022470  F1@10,25,50  [95.09433962 95.09433962 84.52830189]

- average results
Acc: 82.1172  Edit: 90.686691  F1@10,25,50  [93.33801352106744, 92.05041843207468, 84.23859438417979]

Generally I think this reproduciton proves the performance presented in the paper. However I find it difficult to choose which one is better in each split **(Question 2)**, for instance,

> split4 epoch-90.model
> Acc: 83.7560  Edit: 91.022470  F1@10,25,50  [95.09433962 95.09433962 84.52830189]
> split4 epoch-70.model
> Acc: 83.8013  Edit: 92.857870  F1@10,25,50  [95.8490566  95.8490566  83.77358491]

Apparently epoch-90.model gave a better F1@50 while epoch-70.model achieved higher Edit.

I hope you're so kind to clear my two doubts.

**Comment by yo3nglau** [comment]
> Hi, thanks for your recognition of our work! Our training procedure for action segmentation is exactly the same as you have described, and we really appreciate the very detailed explanation you have given us (which can also be a useful complement to our repo).
> 
> We noticed that you trained the ASFormer for 120 epochs, and took the last epoch for evaluation. Here's the problem, we've found that Br-Prompt pre-trained ViT features (d=768) are more effective than the I3D features (d=2048) which were used in the original ASFormer settings. Thus, it may take fewer training epochs for ASFormer to converge. Then we decided to test the results on earlier epochs, and found that the performances could be higher. We're not sure if you used the same random settings to train the framework as we used for the paper, but you can test the model on earlier epochs to verify if there are better results.

Thank you for quick reply! I immediately make efforts based on your suggestion. I met two puzzles. **(Question 1)** If I obtain the best results with certain epoch model for each of 4 splits, but these 4 models might derive from different epochs, can I average these partial best results to acquire the last and of course best performance? I assume that is allowed and the results are shown below:

- split1 epoch-110.model
Acc: 82.4479  Edit: 90.708101  F1@10,25,50  [93.38235294 91.17647059 81.61764706]

- split2 epoch-100.model
Acc: 80.7248  Edit: 87.533241  F1@10,25,50  [89.7810219  87.59124088 81.75182482]

- split3 epoch-100.model
Acc: 81.5399  Edit: 93.482953  F1@10,25,50  [95.09433962 94.33962264 89.05660377]

- split4 epoch-90.model
Acc: 83.7560  Edit: 91.022470  F1@10,25,50  [95.09433962 95.09433962 84.52830189]

- average results
Acc: 82.1172  Edit: 90.686691  F1@10,25,50  [93.33801352106744, 92.05041843207468, 84.23859438417979]

Generally I think this reproduction proves the performance presented in the paper. However I find it difficult to choose which one is better in each split **(Question 2)**, for instance,

> split4 epoch-90.model
> Acc: 83.7560  Edit: 91.022470  F1@10,25,50  [95.09433962 95.09433962 84.52830189]
> split4 epoch-70.model
> Acc: 83.8013  Edit: 92.857870  F1@10,25,50  [95.8490566  95.8490566  83.77358491]

Apparently epoch-90.model gave a better F1@50 while epoch-70.model achieved higher Edit.

I hope you're so kind to clear my two doubts.

**Comment by ttlmh** [comment]
> > Hi, thanks for your recognition of our work! Our training procedure for action segmentation is exactly the same as you have described, and we really appreciate the very detailed explanation you have given us (which can also be a useful complement to our repo).
> > We noticed that you trained the ASFormer for 120 epochs, and took the last epoch for evaluation. Here's the problem, we've found that Br-Prompt pre-trained ViT features (d=768) are more effective than the I3D features (d=2048) which were used in the original ASFormer settings. Thus, it may take fewer training epochs for ASFormer to converge. Then we decided to test the results on earlier epochs, and found that the performances could be higher. We're not sure if you used the same random settings to train the framework as we used for the paper, but you can test the model on earlier epochs to verify if there are better results.
> 
> Thank you for quick reply! I immediately make efforts based on your suggestion. I met two puzzles. **(Question 1)** If I obtain the best results with certain epoch model for each of 4 splits, but these 4 models might derive from different epochs, can I average these partial best results to acquire the last and of course best performance? I assume that is allowed and the results are shown below:
> 
> * split1 epoch-110.model
>   Acc: 82.4479  Edit: 90.708101  F1@10,25,50  [93.38235294 91.17647059 81.61764706]
> * split2 epoch-100.model
>   Acc: 80.7248  Edit: 87.533241  F1@10,25,50  [89.7810219  87.59124088 81.75182482]
> * split3 epoch-100.model
>   Acc: 81.5399  Edit: 93.482953  F1@10,25,50  [95.09433962 94.33962264 89.05660377]
> * split4 epoch-90.model
>   Acc: 83.7560  Edit: 91.022470  F1@10,25,50  [95.09433962 95.09433962 84.52830189]
> * average results
>   Acc: 82.1172  Edit: 90.686691  F1@10,25,50  [93.33801352106744, 92.05041843207468, 84.23859438417979]
> 
> Generally I think this reproduction proves the performance presented in the paper. However I find it difficult to choose which one is better in each split **(Question 2)**, for instance,
> 
> > split4 epoch-90.model
> > Acc: 83.7560  Edit: 91.022470  F1@10,25,50  [95.09433962 95.09433962 84.52830189]
> > split4 epoch-70.model
> > Acc: 83.8013  Edit: 92.857870  F1@10,25,50  [95.8490566  95.8490566  83.77358491]
> 
> Apparently epoch-90.model gave a better F1@50 while epoch-70.model achieved higher Edit.
> 
> I hope you're so kind to clear my two doubts.

For your first doubt, we did average the best results of the four splits to get the final performance. We think this is rational since the dataset sizes of the different splits are different (even if the quantities of video are equal, the video lengths can be varied). Thus, the convergence process should also be distinct.

For your second doubt, we have also found situations where different evaluation metrics respectively receive higher scores on different epochs. We simply choose one of the evaluation metrics (e.g. F1@50) as the standard reference since there is no evidence to suggest which metric is more suitable than the others.

**Comment by yo3nglau** [comment]
You solved all my current doubts within this issue !

Thank you again for your work and guidance!
